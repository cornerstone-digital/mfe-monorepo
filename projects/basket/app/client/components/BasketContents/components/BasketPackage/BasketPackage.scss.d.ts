export type Styles = {
  container: string
  'fade-in': string
  'fade-out': string
  'package-body': string
  'package-body-closed': string
  'package-header': string
  'package-header-overline': string
  'package-header-underline': string
}

export type ClassNames = keyof Styles

declare const styles: Styles

export default styles
