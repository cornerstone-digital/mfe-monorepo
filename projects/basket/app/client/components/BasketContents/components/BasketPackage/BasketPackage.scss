@import 'node_modules/@vfuk/web-core/resources/resources.scss';

.package-header {
  padding: $spacing-fixed-1 $spacing-fixed-2;
}

.package-header-underline {
  border-bottom: 1px solid $color-pale-grey;
}

.package-header-overline {
  border-top: 1px solid $color-pale-grey;
}

.package-body {
  max-height: 10000px; // Required for transition
  transition: max-height 500ms ease-in-out 500ms, opacity 500ms;
  overflow-y: hidden;
  opacity: 1;

  > ul {
    padding-inline-start: 0;
    margin-bottom: 0 !important;
  }
}

.package-body-closed {
  max-height: 0;
  opacity: 0;
}

.container {
  padding: $spacing-fixed-2;
  border-top: 1px solid $color-pale-grey;
}
