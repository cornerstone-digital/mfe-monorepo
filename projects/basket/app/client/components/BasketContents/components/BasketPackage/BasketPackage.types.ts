import { HeaderStatusType } from './components/PackageItemList/PackageItemList.types'

export interface BasketPackageProps {
  thisPackage: PackageWithHeaderStatus | PackageWithHeaderStatus[]
  showPackageHeaderUnderline: boolean
  showBorder: boolean
  phonePaired?: string
  matchedWatchPackages?: PackageWithHeaderStatus[]
  index: number
}

export interface PackageProps extends PackageWithHeaderStatus {
  changePackageLink?: string
  onRemove?: () => void
  onUndo?: () => void
  headerStatus?: HeaderStatusType
  packageId?: BasketV2.ModelPackage['packageId']
}

export interface PackageWithHeaderStatus extends BasketV2.ModelPackage {
  headerStatus?: HeaderStatusType
  tradeInHeaderStatus?: HeaderStatusType
  contractOptions?: BasketV2.Hardware['contractOptions']
}

export interface AddonWithHeaderStatus extends BasketV2.Service {
  headerStatus?: HeaderStatusType
}

export interface BasketParams {
  basketId?: BasketV2.Basket['basketId']
  packageId?: BasketV2.ModelPackage['packageId']
  bundleId?: BasketV2.Bundle['skuId']
  deviceId?: BasketV2.Hardware['skuId']
  extrasId?: BasketV2.Service['skuId'][]
  packageType: BasketV2.ModelPackage['packageType']
  accountCategory: BasketV2.ModelPackage['accountCategory']
  isUpgrade: boolean
  contractOptions: BasketV2.Hardware['contractOptions']
  confirmRequired?: boolean
  tradeInCredit?: BasketV2.TradeInCredit
}

export interface PriceDetails {
  upfrontPrice: string
  monthlyPrice: string
}

export interface BorderDetails {
  borderAppearance?: string
  borderShadow?: boolean
}
