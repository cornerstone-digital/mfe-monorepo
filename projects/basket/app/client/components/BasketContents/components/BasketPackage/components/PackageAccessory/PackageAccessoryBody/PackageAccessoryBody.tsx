import BasketAccessory from '@components/BasketAccessory'
import { isAccessory } from '@helpers/typeCheck'
import ListGroup from '@web-core/components/molecules/ListGroup'
import { PackageAccessoryBodyProps } from './PackageAccessoryBody.types'

const PackageAccessoryBody = (props: PackageAccessoryBodyProps) => {
  const { primaryPackage, isBusiness, onRemoveAddOn, onUndoRemoveAddOn } = props

  const filteredAccessories = primaryPackage.hardwares?.filter(item => isAccessory(item))
  if (!filteredAccessories?.length) {
    return null
  }

  return (
    <ListGroup
      items={filteredAccessories.map(accessory => {
        if (accessory)
          return (
            <BasketAccessory
              key={accessory.packageLineId}
              accessory={accessory}
              packageId={primaryPackage.packageId}
              isBusiness={isBusiness}
              onRemoveAddOn={onRemoveAddOn}
              onUndoRemoveAddOn={onUndoRemoveAddOn}
            />
          )
      })}
    />
  )
}

export default PackageAccessoryBody
