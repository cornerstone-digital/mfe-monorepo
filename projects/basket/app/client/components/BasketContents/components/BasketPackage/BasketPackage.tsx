import { Fragment } from 'react'
import classnames from 'classnames'

import Section from '@web-core/components/atoms/Section'

import BasketItem from '@components/BasketItem'
import PackageFooter from '@components/PackageFooter'
import PackageDiscounts from './components/PackageDiscounts'
import PackageTradeInCredit from './components/PackageTradeInCredit'
import PackageItemList from './components/PackageItemList'

import {
  hasWatch,
  isBusiness as isBusinessType,
  isSimo,
  hasBasketAnyDiscount,
  isAccessoryPackage,
  isAccessory,
  isRedHybridPackage,
} from '@helpers/typeCheck'

import { BasketPackageProps, PackageWithHeaderStatus, BorderDetails } from './BasketPackage.types'

import styles from './BasketPackage.scss'
import getSimoPlanLink from '@helpers/getSimoPlanLink'

import getPackageWithHandset from './helpers/getPackageWithHandset'
import isBingoPackage from './helpers/isBingoPackage'
import getAllActivePackages from './helpers/getAllActivePackages'
import getPackagePrices from './helpers/getPackagePrices'
import getCombinedPackageTotals from './helpers/getCombinedPackageTotals'
import getPairedWatchSubTitle from './helpers/getPairedWatchSubTitle'
import getPairedWatchTitle from './helpers/getPairedWatchTitle'

import BasketPackageContext from './context'
import { useStore } from '@store/index'
import { observer } from 'mobx-react-lite'
import { flowResult } from 'mobx'
import getBasketPackageProps from './helpers/getBasketPackageProps/getBasketPackageProps'
import PackageHeader from './components/PackageHeader/PackageHeader'
import PackageAccessoryBody from './components/PackageAccessory/PackageAccessoryBody'
import PackageAccessoryHeader from './components/PackageAccessory/PackageAccessoryHeader'
import { BasketHardwareWithHeaderStatus } from '@pages/BasketPage/BasketPage.types'

const BasketPackage = observer((props: BasketPackageProps) => {
  const { thisPackage, showBorder, matchedWatchPackages = [], index, showPackageHeaderUnderline, phonePaired } = props
  const allActivePackages = getAllActivePackages(thisPackage, matchedWatchPackages)
  const { basketStore } = useStore()
  const { pageContent, basket, pageError, flags } = basketStore
  const { isUpgradeOrder = false, isReviewMode, isSmallBusiness: isSmallBusinessStore = false, packages } = basket

  const checkOnlyAccessoryPackageLeftOnBasket = (activePackage: PackageWithHeaderStatus) => {
    return (
      isAccessoryPackage(activePackage) &&
      !packages?.filter((item: PackageWithHeaderStatus) => item.packageId !== activePackage.packageId && item.headerStatus !== 'removed')
        .length &&
      activePackage?.hardwares &&
      activePackage.hardwares.filter((h: BasketHardwareWithHeaderStatus) => h.headerStatus !== 'removed').length < 2
    )
  }

  const primaryPackage: PackageWithHeaderStatus = Array.isArray(thisPackage) ? getPackageWithHandset(thisPackage) : thisPackage

  const onUpdateBasket = flowResult((...args: Parameters<typeof basketStore.updateBasket>) => basketStore.updateBasket(...args))
  const onRemoveAddOn = flowResult(
    checkOnlyAccessoryPackageLeftOnBasket(primaryPackage)
      ? (...args: Parameters<typeof basketStore.onRemovePackage>) => basketStore.onRemovePackage(...args)
      : (...args: Parameters<typeof basketStore.onRemoveAddOn>) => basketStore.onRemoveAddOn(...args),
  )

  const onUndoRemoveAddOn = flowResult((...args: Parameters<typeof basketStore.onUndoRemoveAddOn>) =>
    basketStore.onUndoRemoveAddOn(...args),
  )
  const updateStatus = flowResult((...args: Parameters<typeof basketStore.updateStatus>) => basketStore.updateStatus(...args))

  // All active packages that could be included in the 'logical package'. E.G. Handset/SIMO + BB + Watches

  const isBusiness: boolean = isBusinessType(primaryPackage?.accountCategory)
  const isSmallBusiness: boolean = primaryPackage?.accountSubCategory === 'smallBusiness' || isSmallBusinessStore
  const netGrossType: keyof BasketV2.Price = isBusiness ? 'net' : 'gross'
  const accountCategory: string = primaryPackage?.accountCategory || ''
  const primaryPackageId: string | undefined = primaryPackage?.packageId

  const borderColor = isUpgradeOrder ? 'marketingTurquoise' : 'none'
  const borderDetails: BorderDetails = showBorder ? { borderAppearance: borderColor, borderShadow: true } : {}

  const watchPackageIds: PackageWithHeaderStatus['packageId'][] = matchedWatchPackages.map(({ packageId }) => packageId)

  const getPackageProps = (packageWithHeaderStatus: PackageWithHeaderStatus) => {
    return getBasketPackageProps(packageWithHeaderStatus, accountCategory, basketStore, primaryPackageId, watchPackageIds)
  }

  const renderPackageBody = (
    renderedPackage: PackageWithHeaderStatus | PackageWithHeaderStatus[],
    isPairedWatch: boolean,
  ): JSX.Element | JSX.Element[] => {
    const flagsWithBingo = {
      ...flags,
      isBingoEnabled: isBingoPackage(renderedPackage),
    }

    let packageProps = {}
    let combiProps
    let packageTotals

    if (Array.isArray(renderedPackage)) {
      combiProps = renderedPackage.map((combiPackage: PackageWithHeaderStatus) => getPackageProps(combiPackage))
      packageTotals = getCombinedPackageTotals(renderedPackage, false, netGrossType)
    } else {
      packageProps = getPackageProps(renderedPackage)
      packageTotals = getPackagePrices(renderedPackage, false, netGrossType)
    }

    return isAccessoryPackage(primaryPackage) ? (
      <PackageAccessoryBody
        primaryPackage={primaryPackage}
        isBusiness={isBusiness}
        onRemoveAddOn={onRemoveAddOn}
        onUndoRemoveAddOn={onUndoRemoveAddOn}
      />
    ) : (
      <PackageItemList
        onUpdateBasket={onUpdateBasket}
        modelPackage={renderedPackage}
        combiProps={combiProps}
        reviewMode={isReviewMode}
        isBusiness={isBusiness}
        phonePaired={phonePaired}
        hasWatchPackage={!!matchedWatchPackages.length}
        isPairedWatch={isPairedWatch}
        pageError={pageError}
        pageContent={pageContent}
        flags={flagsWithBingo}
        updateStatus={updateStatus}
        onRemoveAddOn={onRemoveAddOn}
        onUndoRemoveAddOn={onUndoRemoveAddOn}
        isUpgrade={isUpgradeOrder}
        {...packageTotals}
        {...packageProps}
      />
    )
  }

  const renderDiscounts = (hasTradeIn: boolean): JSX.Element => {
    return (
      <Fragment key={`${primaryPackageId}-discounts`}>
        <PackageDiscounts
          modelPackage={allActivePackages}
          isBusiness={isBusiness}
          isUpgrade={isUpgradeOrder}
          pageContent={pageContent}
          reviewMode={isReviewMode}
          hasTradeIn={hasTradeIn}
        />
      </Fragment>
    )
  }

  const renderTradeIn = (hasAnyDiscount: boolean): JSX.Element => {
    const tradeInCredit = allActivePackages.find(p => p.tradeInCredit)?.tradeInCredit
    const currentPackage = Array.isArray(thisPackage) ? getPackageWithHandset(thisPackage) : thisPackage

    const handleRemovePackageTradeIn = () => {
      basketStore.onRemovePackageTradeIn(currentPackage.packageId)
    }

    const handleUndoRemovePackageTradeIn = () => {
      basketStore.onUndoRemovePackageTradeIn(currentPackage.packageId)
    }

    return (
      <Fragment key={`${primaryPackageId}-tradeInCredit`}>
        {tradeInCredit && currentPackage?.packageId && (
          <PackageTradeInCredit
            packageId={currentPackage.packageId}
            pageContent={pageContent}
            tradeInCredit={tradeInCredit}
            headerStatus={currentPackage.tradeInHeaderStatus}
            hideTopBorder={hasAnyDiscount && currentPackage.tradeInHeaderStatus !== 'removed'}
            onRemove={handleRemovePackageTradeIn}
            onUndo={handleUndoRemovePackageTradeIn}
          />
        )}
      </Fragment>
    )
  }

  const renderPackageTotal = (): JSX.Element => {
    const { upfrontPrice, monthlyPrice } = getCombinedPackageTotals(allActivePackages, true, netGrossType)

    return (
      <div key={`${primaryPackageId}-totals`} className={styles['container']}>
        <BasketItem upfrontPrice={upfrontPrice} monthlyPrice={monthlyPrice} title={'Package Total'} isBusiness={isBusiness} isTotalCost />
      </div>
    )
  }

  const renderPackageHeader = (
    renderedPackage: PackageWithHeaderStatus | PackageWithHeaderStatus[],
    isPairedWatch: boolean,
  ): JSX.Element => {
    let [title, subTitle] =
      isPairedWatch && !Array.isArray(renderedPackage)
        ? [getPairedWatchTitle(renderedPackage), getPairedWatchSubTitle(renderedPackage)]
        : [`Package ${index + 1}`]

    if (isRedHybridPackage(thisPackage as BasketV2.ModelPackage)) {
      title = 'Flexi Plan'
      subTitle = ''
    }

    return isAccessoryPackage(primaryPackage) ? (
      <PackageAccessoryHeader />
    ) : (
      <PackageHeader
        renderedPackage={renderedPackage}
        primaryPackage={primaryPackage}
        isPairedWatch={isPairedWatch}
        title={title}
        subTitle={subTitle}
        showPackageHeaderUnderline={showPackageHeaderUnderline}
      />
    )
  }

  const renderJourneyLinkField = () => {
    const mainPackages = Array.isArray(thisPackage) ? thisPackage : [thisPackage]
    const simoPackage = mainPackages.find(pkg => isSimo(pkg.bundle))

    if (simoPackage && isUpgradeOrder) {
      return <input type="hidden" id="journeyId" value={getSimoPlanLink(simoPackage)} />
    }
    return null
  }

  const hasActivePackage = (p: PackageWithHeaderStatus | PackageWithHeaderStatus[]): boolean => {
    if (Array.isArray(p)) {
      return !!p.filter(hasActivePackage).length
    }

    return isAccessory(p) || p?.headerStatus !== 'removed'
  }

  const renderAllPackages = (): JSX.Element[] => {
    // Different to allActivePackages because this passes through combi packages as and array for rendering
    const packagesToRender: (PackageWithHeaderStatus | PackageWithHeaderStatus[])[] = [thisPackage, ...matchedWatchPackages]
    const maxIndex = packagesToRender.length - 1
    const activePackageCount = packagesToRender.filter(hasActivePackage).length
    const hasAnyDiscount = hasBasketAnyDiscount(allActivePackages)
    const hasTradeIn = allActivePackages.some(p => !!p.tradeInCredit)
    const isCombinedPackage = allActivePackages.length > 1
    const showPackageTotal = hasAnyDiscount || isCombinedPackage
    return packagesToRender.map((singlePackage: PackageWithHeaderStatus | PackageWithHeaderStatus[], pIndex: number) => {
      const isRemoved = !hasActivePackage(singlePackage)
      const packageHasWatch = !Array.isArray(singlePackage) && hasWatch(singlePackage?.hardwares)
      const isNotAccessoryPackage = !isAccessoryPackage(singlePackage as BasketV2.ModelPackage)
      // classnames for the full body when there are no visible packages remaining.
      // !! If there are no packages remaining within the 'logical package' then we also remove the discounts and package total
      const bodyClassNames = classnames('package-body', {
        'package-body-closed': isRemoved && !activePackageCount && isNotAccessoryPackage,
      })
      // classnames specifically for the package, excluding discounts and total.
      // !! This is so that we can animate the package removal without affecting the discounts and package total
      const packageBodyClassNames = classnames('package-body', {
        'package-body-closed': isRemoved && isNotAccessoryPackage,
      })

      return (
        <BasketPackageContext.Provider key={pIndex} value={{ ...props, isSmallBusiness }}>
          {renderPackageHeader(singlePackage, packageHasWatch)}
          <div styleName={bodyClassNames}>
            <div styleName={packageBodyClassNames}>{renderPackageBody(singlePackage, packageHasWatch)}</div>
            <If condition={pIndex === maxIndex}>
              <If condition={hasAnyDiscount}>{renderDiscounts(hasTradeIn)}</If>
              <If condition={hasTradeIn}>{renderTradeIn(hasAnyDiscount)}</If>
              <If condition={!isSmallBusiness && showPackageTotal}>{renderPackageTotal()}</If>
            </If>
            <PackageFooter package={singlePackage} pageContent={pageContent} />
          </div>
        </BasketPackageContext.Provider>
      )
    })
  }

  return (
    <Section marginVertical={2} {...borderDetails}>
      {renderJourneyLinkField()}
      {renderAllPackages()}
    </Section>
  )
})

BasketPackage.displayName = 'BasketPackage'

export default BasketPackage
