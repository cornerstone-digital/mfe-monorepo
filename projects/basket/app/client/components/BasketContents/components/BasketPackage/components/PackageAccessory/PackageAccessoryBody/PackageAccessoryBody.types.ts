import { PackageWithHeaderStatus } from '../../../BasketPackage.types'

export interface PackageAccessoryBodyProps {
  primaryPackage: PackageWithHeaderStatus
  isBusiness: boolean
  onRemoveAddOn: never
  onUndoRemoveAddOn: never
}
