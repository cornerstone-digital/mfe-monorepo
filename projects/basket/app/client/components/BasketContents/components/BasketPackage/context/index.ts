import { createContext } from 'react'
import { BasketPackageProps } from '../BasketPackage.types'

export type BasketPackageContextProps = Partial<BasketPackageProps> & { isSmallBusiness?: boolean }

const BasketPackageContext = createContext<BasketPackageContextProps>({})

export default BasketPackageContext
