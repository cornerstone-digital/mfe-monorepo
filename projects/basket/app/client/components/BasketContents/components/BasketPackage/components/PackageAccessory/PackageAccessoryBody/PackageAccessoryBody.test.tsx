import { render } from '@testing-library/react'
import PackageAccessoryBody from './PackageAccessoryBody'
import { TransformedBasket } from '@shared/helpers/basketTransformer/BasketTransformer'
import { BasketPackageProps, PackageWithHeaderStatus } from '../../../BasketPackage.types'

import accessoryBasketPackageMock from '../mocks/basket1.mock.json'

const accessoryBasketPackage: BasketPackageProps = TransformUtils.removeNulls(accessoryBasketPackageMock)

import pageContent from '@shared/config/content/BasketPageContent.json'
import TransformUtils from '@shared/helpers/transformUtils/transformUtils'

const basketPageContent: BasketPageContent.Basket = pageContent[0]?.basket as BasketPageContent.Basket

function mockStore(basket: Partial<TransformedBasket> = {}): any {
  return {
    basketStore: {
      basket: {
        basketId: '123',
        ...basket,
      },
      pageContent: basketPageContent,
    },
  }
}

jest.mock('@store', () => ({
  useStore: jest.fn().mockImplementation(() => mockStore()),
}))

describe('PackageAccessoryBody', () => {
  it('should render null', () => {
    const primaryPackage = { ...(accessoryBasketPackage.thisPackage as PackageWithHeaderStatus) }
    primaryPackage.hardwares = []
    const { container } = render(
      <PackageAccessoryBody
        primaryPackage={primaryPackage}
        isBusiness={false}
        onRemoveAddOn={jest.fn() as never}
        onUndoRemoveAddOn={jest.fn() as never}
      />,
    )
    expect(container.firstChild).toBeNull()
  })

  it('should render for basic scenario', () => {
    const { getByText, queryAllByText, queryByText, asFragment } = render(
      <PackageAccessoryBody
        primaryPackage={accessoryBasketPackage.thisPackage as PackageWithHeaderStatus}
        isBusiness={false}
        onRemoveAddOn={jest.fn() as never}
        onUndoRemoveAddOn={jest.fn() as never}
      />,
    )

    expect(getByText('XQISIT Wireless Charger 10W Black')).toBeInTheDocument()
    expect(queryAllByText('Remove')).toHaveLength(3)
    expect(queryByText('undefined-month Airtime Plan')).not.toBeInTheDocument()
    expect(queryByText('Joining us from another network')).not.toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
