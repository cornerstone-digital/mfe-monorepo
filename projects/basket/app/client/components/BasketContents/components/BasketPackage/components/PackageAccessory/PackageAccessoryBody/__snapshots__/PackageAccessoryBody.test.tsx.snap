// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageAccessoryBody should render for basic scenario 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="vfuk-ListGroup__list vfuk-ListGroup__line"
    >
      <div
        class="sc-bdfBQB Gridstyle__GridContainer-sc-11onwkc-0 juDKGU bRCyJL no-gutters"
        data-name="container-fluid"
      >
        <div
          class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt"
          data-name="row"
        >
          <div
            class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 bVSFOy cRslqE header"
            data-name="col-12 col-md-12 col-lg-12"
          >
            <div
              stylename="item-header"
            >
              <div
                class="item-title"
              >
                <h5
                  class="heading heading-6 inherit w-bold margin-0"
                >
                  XQISIT Wireless Charger 10W Black
                </h5>
              </div>
              <button
                data-selector="basket-header-action-xqisit-wireless-charger-10w-black"
                stylename="action-button"
              >
                <span
                  class="action-text"
                >
                  Remove
                </span>
                <span
                  aria-hidden="true"
                  class="wrapper inherit size-1 1  "
                >
                  test-file-stub
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          aria-hidden="false"
          class="item-body-wrap"
        >
          <div
            class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt item-body"
            data-name="row"
          >
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 heDFIj cRslqE icon-column-wrapper"
              data-name="col-4 col-md-2 col-lg-2"
            >
              <picture
                class="product-image vfuk-image__picture "
              >
                <source
                  media="(min-width: screenDesktop)"
                  srcset="http://mock-localhost/images/desktop/accessory_sku200913_black-grid-product-front.png"
                />
                <source
                  media="(min-width: screenTablet)"
                />
                <source
                  media="(min-width: 0)"
                  srcset="http://mock-localhost/images/desktop/accessory_sku200913_black-grid-product-front.png"
                />
                <img
                  alt="XQISIT Wireless Charger 10W Black"
                  height="auto"
                  src="http://mock-localhost/images/desktop/accessory_sku200913_black-grid-product-front.png"
                  width="auto"
                />
              </picture>
            </div>
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 fVIZqk cRslqE justify-left desc-column-large"
              data-name="col-12 col-md-5 col-lg-4"
            />
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 ebiDZn cRslqE"
              data-name="col-8 col-md-5 col-lg-6"
            >
              <div
                class="sc-bdfBQB Gridstyle__GridContainer-sc-11onwkc-0 juDKGU bRCyJL no-gutters"
                data-name="container-fluid"
              >
                <div
                  class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt"
                  data-name="row"
                >
                  <div
                    class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 dOTIbJ cRslqE column"
                    data-name="col-6"
                  >
                    <div
                      class="basket-item-totals"
                      data-selector="cost-upfront"
                    >
                      <p
                        class=""
                      >
                        <span
                          class="prefix inherit"
                        >
                          Upfront
                        </span>
                        <span
                          class="main inherit"
                        >
                          £15
                        </span>
                        <span
                          class="suffix inherit"
                        />
                        <span
                          class="suffix2 grey"
                        />
                      </p>
                    </div>
                  </div>
                  <div
                    class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 dOTIbJ cRslqE column"
                    data-name="col-6"
                  >
                    <div
                      class="basket-item-totals"
                      data-selector="cost-monthly"
                    >
                      <p
                        class=""
                      >
                        <span
                          class="prefix inherit"
                        >
                          Monthly
                        </span>
                        <span
                          class="main inherit"
                        >
                          £0
                        </span>
                        <span
                          class="suffix inherit"
                        />
                        <span
                          class="suffix2 grey"
                        />
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      class="vfuk-ListGroup__list vfuk-ListGroup__line"
    >
      <div
        class="sc-bdfBQB Gridstyle__GridContainer-sc-11onwkc-0 juDKGU bRCyJL no-gutters"
        data-name="container-fluid"
      >
        <div
          class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt"
          data-name="row"
        >
          <div
            class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 bVSFOy cRslqE header"
            data-name="col-12 col-md-12 col-lg-12"
          >
            <div
              stylename="item-header"
            >
              <div
                class="item-title"
              >
                <h5
                  class="heading heading-6 inherit w-bold margin-0"
                >
                  Apple EarPods Lightning White
                </h5>
              </div>
              <button
                data-selector="basket-header-action-apple-earpods-lightning-white"
                stylename="action-button"
              >
                <span
                  class="action-text"
                >
                  Remove
                </span>
                <span
                  aria-hidden="true"
                  class="wrapper inherit size-1 1  "
                >
                  test-file-stub
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          aria-hidden="false"
          class="item-body-wrap"
        >
          <div
            class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt item-body"
            data-name="row"
          >
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 heDFIj cRslqE icon-column-wrapper"
              data-name="col-4 col-md-2 col-lg-2"
            >
              <picture
                class="product-image vfuk-image__picture "
              >
                <source
                  media="(min-width: screenDesktop)"
                  srcset="http://mock-localhost/images/desktop/accessory_sku093337_white-grid-product-front.png"
                />
                <source
                  media="(min-width: screenTablet)"
                />
                <source
                  media="(min-width: 0)"
                  srcset="http://mock-localhost/images/desktop/accessory_sku093337_white-grid-product-front.png"
                />
                <img
                  alt="Apple EarPods Lightning White"
                  height="auto"
                  src="http://mock-localhost/images/desktop/accessory_sku093337_white-grid-product-front.png"
                  width="auto"
                />
              </picture>
            </div>
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 fVIZqk cRslqE justify-left desc-column-large"
              data-name="col-12 col-md-5 col-lg-4"
            />
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 ebiDZn cRslqE"
              data-name="col-8 col-md-5 col-lg-6"
            >
              <div
                class="sc-bdfBQB Gridstyle__GridContainer-sc-11onwkc-0 juDKGU bRCyJL no-gutters"
                data-name="container-fluid"
              >
                <div
                  class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt"
                  data-name="row"
                >
                  <div
                    class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 dOTIbJ cRslqE column"
                    data-name="col-6"
                  >
                    <div
                      class="basket-item-totals"
                      data-selector="cost-upfront"
                    >
                      <p
                        class=""
                      >
                        <span
                          class="prefix inherit"
                        >
                          Upfront
                        </span>
                        <span
                          class="main inherit"
                        >
                          £19
                        </span>
                        <span
                          class="suffix inherit"
                        />
                        <span
                          class="suffix2 grey"
                        />
                      </p>
                    </div>
                  </div>
                  <div
                    class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 dOTIbJ cRslqE column"
                    data-name="col-6"
                  >
                    <div
                      class="basket-item-totals"
                      data-selector="cost-monthly"
                    >
                      <p
                        class=""
                      >
                        <span
                          class="prefix inherit"
                        >
                          Monthly
                        </span>
                        <span
                          class="main inherit"
                        >
                          £0
                        </span>
                        <span
                          class="suffix inherit"
                        />
                        <span
                          class="suffix2 grey"
                        />
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      class="vfuk-ListGroup__list vfuk-ListGroup__line"
    >
      <div
        class="sc-bdfBQB Gridstyle__GridContainer-sc-11onwkc-0 juDKGU bRCyJL no-gutters"
        data-name="container-fluid"
      >
        <div
          class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt"
          data-name="row"
        >
          <div
            class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 bVSFOy cRslqE header"
            data-name="col-12 col-md-12 col-lg-12"
          >
            <div
              stylename="item-header"
            >
              <div
                class="item-title"
              >
                <h5
                  class="heading heading-6 inherit w-bold margin-0"
                >
                  InvisibleShield Glass Elite+ Clear
                </h5>
              </div>
              <button
                data-selector="basket-header-action-invisibleshield-glass-elite-clear"
                stylename="action-button"
              >
                <span
                  class="action-text"
                >
                  Remove
                </span>
                <span
                  aria-hidden="true"
                  class="wrapper inherit size-1 1  "
                >
                  test-file-stub
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          aria-hidden="false"
          class="item-body-wrap"
        >
          <div
            class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt item-body"
            data-name="row"
          >
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 heDFIj cRslqE icon-column-wrapper"
              data-name="col-4 col-md-2 col-lg-2"
            >
              <picture
                class="product-image vfuk-image__picture "
              >
                <source
                  media="(min-width: screenDesktop)"
                  srcset="http://mock-localhost/images/desktop/accessory_sku208431_clear-grid-product-front.png"
                />
                <source
                  media="(min-width: screenTablet)"
                />
                <source
                  media="(min-width: 0)"
                  srcset="http://mock-localhost/images/desktop/accessory_sku208431_clear-grid-product-front.png"
                />
                <img
                  alt="InvisibleShield Glass Elite+ Clear"
                  height="auto"
                  src="http://mock-localhost/images/desktop/accessory_sku208431_clear-grid-product-front.png"
                  width="auto"
                />
              </picture>
            </div>
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 fVIZqk cRslqE justify-left desc-column-large"
              data-name="col-12 col-md-5 col-lg-4"
            />
            <div
              class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 ebiDZn cRslqE"
              data-name="col-8 col-md-5 col-lg-6"
            >
              <div
                class="sc-bdfBQB Gridstyle__GridContainer-sc-11onwkc-0 juDKGU bRCyJL no-gutters"
                data-name="container-fluid"
              >
                <div
                  class="sc-gsTEea GridRowstyle__GridRow-sc-1l0dhq9-0 DvuGS dhUlIt"
                  data-name="row"
                >
                  <div
                    class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 dOTIbJ cRslqE column"
                    data-name="col-6"
                  >
                    <div
                      class="basket-item-totals"
                      data-selector="cost-upfront"
                    >
                      <p
                        class=""
                      >
                        <span
                          class="prefix inherit"
                        >
                          Upfront
                        </span>
                        <span
                          class="main inherit"
                        >
                          £10
                        </span>
                        <span
                          class="suffix inherit"
                        />
                        <span
                          class="suffix2 grey"
                        />
                      </p>
                    </div>
                  </div>
                  <div
                    class="sc-dlfnuX GridColumnstyle__GridColumn-sc-19afl1d-0 dOTIbJ cRslqE column"
                    data-name="col-6"
                  >
                    <div
                      class="basket-item-totals"
                      data-selector="cost-monthly"
                    >
                      <p
                        class=""
                      >
                        <span
                          class="prefix inherit"
                        >
                          Monthly
                        </span>
                        <span
                          class="main inherit"
                        >
                          £0
                        </span>
                        <span
                          class="suffix inherit"
                        />
                        <span
                          class="suffix2 grey"
                        />
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
